src/simd.rs

```diff
@@ -0,0 +1,215 @@
+//! SIMD byte scanning logic.
+//! 
+//! This module provides functions that allow walking through byteslices, calling
+//! provided callback functions on special bytes and their indices using SIMD.
+//! The byteset is defined in `compute_lookup`.
+//!
+//! The idea is to load in a chunk of 16 bytes and perform a lookup into a set of
+//! bytes on all the bytes in this chunk simultaneously. We produce a 16 bit bitmask
+//! from this and call the callback on every index corresponding to a 1 in this mask
+//! before moving on to the next chunk. This allows us to move quickly when there
+//! are no or few matches.
+//!
+//! The table lookup is inspired by this [great overview]. However, since all of the
+//! bytes we're interested in are ASCII, we don't quite need the fully generality of
+//! the universal algorithm and are hence able to skip a few instructions.
+//!
+//! [great overview]: http://0x80.pl/articles/simd-byte-lookup.html
+
+use core::arch::x86_64::*;
+use crate::parse::LoopInstruction;
+
+const VECTOR_SIZE: usize = std::mem::size_of::<__m128i>();
+
+/// Generates a lookup table containing the bitmaps for our
+/// special marker bytes. This is effectively a 128 element 2d bitvector,
+/// that can be indexed by a four bit row index (the lower nibble)
+/// and a three bit column index (upper nibble).
+const fn compute_lookup() -> [u8; 16] {
```

Comment:
Just wanna say, using a const fn for this kind of thing is great, much better than having random packed hex values in the code - in particular, easy to see how to update in case of more markup characters.